plugins/catalog-backend/src/ingestion/processors/GitlabReaderProcessor.ts

```diff
@@ -15,36 +15,252 @@
  */
 
 import { LocationSpec } from '@backstage/catalog-model';
-import fetch from 'node-fetch';
+import { Config } from '@backstage/config';
+import parseGitUri from 'git-url-parse';
+import fetch, { HeadersInit, RequestInit } from 'node-fetch';
+import { Logger } from 'winston';
 import * as result from './results';
 import { LocationProcessor, LocationProcessorEmit } from './types';
 
-export class GitlabReaderProcessor implements LocationProcessor {
+/**
+ * The configuration parameters for a single GitLab API provider.
+ */
+export type ProviderConfig = {
+  /**
+   * The prefix of the target that this matches on, e.g. "https://gitlab.com",
+   * with no trailing slash.
+   */
+  target: string;
+
+  /**
+   * The base URL of the API of this provider, e.g. "https://gitlab.com/api/v4",
+   * with no trailing slash.
+   *
+   * May be omitted specifically for GitLab; then it will be deduced.
+   *
+   * The API will always be preferred if both its base URL and a token are
+   * present.
+   */
+  apiBaseUrl?: string;
+
+  /**
+   * The base URL of the raw fetch endpoint of this provider, e.g.
+   * "https://gitlab.com", with no trailing slash.
+   *
+   * May be omitted specifically for GitLab; then it will be deduced.
+   *
+   * The API will always be preferred if both its base URL and a token are
+   * present.
+   */
+  rawBaseUrl?: string;
+
+  /**
+   * The authorization token to use for requests to this provider.
+   *
+   * If no token is specified, anonymous access is used.
+   */
+  token?: string;
+};
+
+export function getApiRequestOptions(provider: ProviderConfig): RequestInit {
+  const headers: HeadersInit = {
+    Accept: 'application/json',
+  };
+
+  if (provider.token) {
+    headers['PRIVATE-TOKEN'] = provider.token;
+  }
+
+  return {
+    headers,
+  };
+}
+
+export function getRawRequestOptions(provider: ProviderConfig): RequestInit {
+  const headers: HeadersInit = {};
+
+  if (provider.token) {
+    headers['PRIVATE-TOKEN'] = provider.token;
+  }
+
+  return {
+    headers,
+  };
+}
+
+// Converts for example
+// from: https://gitlab.com/a/b/blob/branchname/path/to/c.yaml
+// to:   https://gitlab.com/api/v4/projects/a/b/repository/files/path/to/c.yaml/raw?ref=branchname
+export function getApiUrl(target: string, provider: ProviderConfig): URL {
+  try {
+    const { owner, name, ref, filepathtype, filepath } = parseGitUri(target);
+
+    if (
+      !owner ||
+      !name ||
+      !ref ||
+      (filepathtype !== 'blob' && filepathtype !== 'raw') ||
+      !filepath?.match(/\.ya?ml$/)
```

Comment:
You can remove this line (and the other three similar ones below). These readers are now used in a more general way, e.g. for reading CODEOWNERS files etc.